d2/src/main/java/com/linkedin/d2/xds/XdsClientImpl.java

```diff
@@ -222,15 +222,24 @@ private void handleD2NodeResponse(DiscoveryResponseData data)
       try
       {
         XdsD2.Node d2Node = resource.getResource().unpack(XdsD2.Node.class);
-        updates.put(resourceName, new NodeUpdate(resource.getVersion(), d2Node));
-      } catch (InvalidProtocolBufferException e)
+        if (d2Node != null && d2Node.getData().isEmpty())
```

Comment:
if d2Node.getData() is empty, it stills to be put into updates map, so that the Subscriber for it can be notified, otherwise any event bus listener will time out. Line 404 (the null guard in the Subscriber) should also check if the actual data is empty and avoid overriding the property.

Same applies to the handleD2URIMapResponse.

Reply:
1. Thanks for reminding. Update the logic to avoid bus listener timeout
2. For Line 404. Do u mean ?
```
if (data != null)
      { // null value guard to avoid overwriting the property with null
        _data = data;
      }
```

Reply:
yeah, it should be 
```
if (data != null && CHECK_ACTUAL_DATA_INSIDE_IS_NOT_EMPTY)
{
  _data = data
}

Reply:
Updated. Add empty check logic for data